projects/kbd-2.6.4/docs/man/man1/loadkeys.1.in
.\" @(#)loadkeys.1
.TH LOADKEYS 1 "6 Feb 1994" "kbd"
.SH NAME
loadkeys \- load keyboard translation tables
.SH SYNOPSIS
.B loadkeys
[\fI\,OPTION\/\fR]... \fI\,FILENAME\/\fR...
.br
.B loadkeys
.I --default
.br
.B loadkeys
.I --mktable
.br
.B loadkeys
.I --bkeymap
.br
.B loadkeys
.I --parse
.LP
.SH DESCRIPTION
.IX "loadkeys command" "" "\fLloadkeys\fR command"
.LP
The program
.B loadkeys
reads the file or files specified by
.IR FILENAME... .
Its main purpose is to load the kernel keymap for the console.
You can specify console device by the
.I -C
(or
.I --console
) option.
.SH "RESET TO DEFAULT"
If the
.I -d
(or
.I --default
) option is given,
.B loadkeys
loads a default keymap, probably the file
.I defkeymap.map
either in
.I @DATADIR@/keymaps
or in
.IR /usr/src/linux/drivers/char .
(Probably the former was user-defined, while the latter
is a qwerty keyboard map for PCs - maybe not what was desired.)
Sometimes, with a strange keymap loaded (with the minus on
some obscure unknown modifier combination) it is easier to
type `loadkeys defkeymap'.
.SH "LOAD KERNEL KEYMAP"
The main function of
.B loadkeys
is to load or modify the keyboard driver's translation tables.
When specifying the file names, standard input can be denoted
by dash (-). If no file is specified, the data is read from
the standard input.
.LP
For many countries and keyboard types appropriate keymaps
are available already, and a command like `loadkeys uk' might
do what you want. On the other hand, it is easy to construct
one's own keymap. The user has to tell what symbols belong
to each key. She can find the keycode for a key by use of
.BR showkey (1),
while the keymap format is given in
.BR keymaps (5)
and can also be seen from the output of
.BR dumpkeys (1).
.SH "LOAD KERNEL ACCENT TABLE"
If the input file does not contain any compose key definitions,
the kernel accent table is left unchanged, unless the
.I -c
(or
.I --clearcompose
) option is given, in which case the kernel accent table is emptied.
If the input file does contain compose key definitions, then all
old definitions are removed, and replaced by the specified new entries.
The kernel accent table is a sequence of (by default 68) entries
describing how dead diacritical signs and compose keys behave.
For example, a line
.LP
.RS
compose ',' 'c' to ccedilla
.RE
.LP
means that <ComposeKey><,><c> must be combined to <ccedilla>.
The current content of this table can be see
using `dumpkeys \-\-compose\-only'.
.SH "LOAD KERNEL STRING TABLE"
The option
.I -s
(or
.I --clearstrings
) clears the kernel string table. If this option is not given,
.B loadkeys
will only add or replace strings, not remove them.
(Thus, the option \-s is required to reach a well-defined state.)
The kernel string table is a sequence of strings
with names like F31. One can make function key F5 (on
an ordinary PC keyboard) produce the text `Hello!',
and Shift+F5 `Goodbye!' using lines
.LP
.RS
keycode 63 = F70 F71
.br
string F70 = "Hello!"
.br
string F71 = "Goodbye!"
.RE
.LP
in the keymap.
The default bindings for the function keys are certain
escape sequences mostly inspired by the VT100 terminal.
.SH "CREATE KERNEL SOURCE TABLE"
If the
.I -m
(or
.I --mktable
) option is given
.B loadkeys
prints to the standard output a file that may be used as
.I /usr/src/linux\%/drivers\%/char\%/defkeymap.c,
specifying the default key bindings for a kernel
(and does not modify the current keymap).
.SH "CREATE BINARY KEYMAP"
If the
.I -b
(or
.I --bkeymap
) option is given
.B loadkeys
prints to the standard output a file that may be used as a binary
keymap as expected by Busybox
.B loadkmap
command (and does not modify the current keymap).
.SH "UNICODE MODE"
.B loadkeys
automatically detects whether the console is in Unicode or
ASCII (XLATE) mode.  When a keymap is loaded, literal
keysyms (such as
.BR section )
are resolved accordingly; numerical keysyms are converted to
fit the current console mode, regardless of the way they are
specified (decimal, octal, hexadecimal or Unicode).
.LP
The
.I -u
(or
.IR --unicode )
switch forces
.B loadkeys
to convert all keymaps to Unicode.  If the keyboard is in a
non-Unicode mode, such as XLATE,
.B loadkeys
will change it to Unicode for the time of its execution.  A
warning message will be printed in this case.
.LP
It is recommended to run
.BR kbd_mode (1)
before
.B loadkeys
instead of using the
.I -u
option.
.SH "OTHER OPTIONS"
.TP
.B \-a \-\-ascii
Force conversion to ASCII.
.TP
.B \-h \-\-help
.B loadkeys
prints its version number and a short usage message to the programs
standard error output and exits.
.TP
.B \-p \-\-parse
.B loadkeys
searches and parses keymap without action.
.TP
.B \-q \-\-quiet
.B loadkeys
suppresses all normal output.
.TP
.B \-V \-\-version
.B loadkeys
prints version number and exits.
.SH WARNING
Note that anyone having read access to
.B /dev/console
can run
.B loadkeys
and thus change the keyboard layout, possibly making it unusable. Note
that the keyboard translation table is common for all the virtual
consoles, so any changes to the keyboard bindings affect all the virtual
consoles simultaneously.
.LP
Note that because the changes affect all the virtual consoles, they also
outlive your session. This means that even at the login prompt the key
bindings may not be what the user expects.
.SH FILES
.TP
.I @DATADIR@/keymaps
default directory for keymaps.
.LP
.TP
.I /usr/src/linux/drivers/char/defkeymap.map
default kernel keymap.
.LP
.SH "SEE ALSO"
.BR dumpkeys (1),
.BR keymaps (5)



projects/kbd-2.6.4/src/loadkeys.c
/* loadkeys.c
 *
 * This file is part of kbd project.
 * Copyright (C) 1993  Risto Kankkunen.
 * Copyright (C) 1993  Eugene G. Crosser.
 * Copyright (C) 1994-2007  Andries E. Brouwer.
 * Copyright (C) 2007-2012  Alexey Gladkov <gladkov.alexey@gmail.com>
 *
 * This file is covered by the GNU General Public License,
 * which should be included with kbd as the file COPYING.
 */
#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <unistd.h>
#include <errno.h>
#include <sysexits.h>
#include <sys/ioctl.h>

#include "libcommon.h"

#include "paths.h"
#include "keymap.h"

static const char *const dirpath1[] = {
	DATADIR "/" KEYMAPDIR "/**",
	KERNDIR "/",
	NULL
};
static const char *const suffixes[] = {
	"",
	".kmap",
	".map",
	NULL
};

static void __attribute__((noreturn))
usage(int rc, const struct kbd_help *options)
{
	const struct kbd_help *h;

	fprintf(stderr, _("Usage: %s [option...] [mapfile...]\n"), get_progname());

	if (options) {
		int max = 0;

		fprintf(stderr, "\n");
		fprintf(stderr, _("Options:"));
		fprintf(stderr, "\n");

		for (h = options; h && h->opts; h++) {
			int len = (int) strlen(h->opts);
			if (max < len)
				max = len;
		}
		max += 2;

		for (h = options; h && h->opts; h++)
			fprintf(stderr, "  %-*s %s\n", max, h->opts, h->desc);
	}

	fprintf(stderr, "\n");
	fprintf(stderr, _("Default keymap: %s\n"), DEFMAP);
	fprintf(stderr, "\n");
	fprintf(stderr, _("Report bugs to authors.\n"));
	fprintf(stderr, "\n");

	exit(rc);
}

int main(int argc, char *argv[])
{
	int options = 0;

	const char *const *dirpath;
	const char *dirpath2[] = { NULL, NULL };

	struct lk_ctx *ctx;
	lk_flags flags = 0;

	int c, i, rc = -1;
	int fd = -1;
	int kbd_mode;
	int kd_mode;
	char *console = NULL;
	char *ev;
	struct kbdfile_ctx *fctx;
	struct kbdfile *fp = NULL;

	set_progname(argv[0]);
	setuplocale();

	const char *const short_opts = "abcC:dhmpsuqvV";
	const struct option long_opts[] = {
		{ "console", required_argument, NULL, 'C' },
		{ "ascii", no_argument, NULL, 'a' },
		{ "bkeymap", no_argument, NULL, 'b' },
		{ "clearcompose", no_argument, NULL, 'c' },
		{ "default", no_argument, NULL, 'd' },
		{ "help", no_argument, NULL, 'h' },
		{ "mktable", no_argument, NULL, 'm' },
		{ "parse", no_argument, NULL, 'p' },
		{ "clearstrings", no_argument, NULL, 's' },
		{ "unicode", no_argument, NULL, 'u' },
		{ "quiet", no_argument, NULL, 'q' },
		{ "verbose", no_argument, NULL, 'v' },
		{ "version", no_argument, NULL, 'V' },
		{ NULL, 0, NULL, 0 }
	};
	const struct kbd_help opthelp[] = {
		{ "-C, --console=DEV",  _("the console device to be used.") },
		{ "-a, --ascii",        _("force conversion to ASCII.") },
		{ "-b, --bkeymap",      _("output a binary keymap to stdout.") },
		{ "-c, --clearcompose", _("clear kernel compose table.") },
		{ "-d, --default",      _("load default.") },
		{ "-m, --mktable",      _("output a 'defkeymap.c' to stdout.") },
		{ "-p, --parse",        _("search and parse keymap without action.") },
		{ "-s, --clearstrings", _("clear kernel string table.") },
		{ "-u, --unicode",      _("force conversion to Unicode.") },
		{ "-q, --quiet",        _("suppress all normal output.") },
		{ "-v, --verbose",      _("be more verbose.") },
		{ "-V, --version",      _("print version number.")     },
		{ "-h, --help",         _("print this usage message.") },
		{ NULL, NULL }
	};

	enum options {
		OPT_A = (1 << 1),
		OPT_B = (1 << 2),
		OPT_D = (1 << 3),
		OPT_M = (1 << 4),
		OPT_U = (1 << 5),
		OPT_P = (1 << 6)
	};

	ctx = lk_init();
	if (!ctx) {
		exit(EXIT_FAILURE);
	}

	if (!(fctx = kbdfile_context_new()))
		kbd_error(EXIT_FAILURE, errno, _("Unable to create kbdfile context"));

	while ((c = getopt_long(argc, argv, short_opts, long_opts, NULL)) != -1) {
		switch (c) {
			case 'a':
				options |= OPT_A;
				break;
			case 'b':
				options |= OPT_B;
				break;
			case 'c':
				flags |= LK_FLAG_CLEAR_COMPOSE;
				break;
			case 'C':
				console = optarg;
				break;
			case 'd':
				options |= OPT_D;
				break;
			case 'm':
				options |= OPT_M;
				break;
			case 'p':
				options |= OPT_P;
				break;
			case 's':
				flags |= LK_FLAG_CLEAR_STRINGS;
				break;
			case 'u':
				options |= OPT_U;
				flags |= LK_FLAG_UNICODE_MODE;
				flags |= LK_FLAG_PREFER_UNICODE;
				break;
			case 'q':
				lk_set_log_priority(ctx, LOG_ERR);
				break;
			case 'v':
				lk_set_log_priority(ctx, LOG_INFO);
				break;
			case 'V':
				print_version_and_exit();
				break;
			case 'h':
				usage(EXIT_SUCCESS, opthelp);
				break;
			case '?':
				usage(EX_USAGE, opthelp);
				break;
		}
	}

	if ((options & OPT_U) && (options & OPT_A))
		kbd_error(EXIT_FAILURE, 0, _("Options %s and %s are mutually exclusive."),
				"--unicode", "--ascii");

	if (!(options & OPT_M) && !(options & OPT_B)) {
		/* get console */
		if ((fd = getfd(console)) < 0)
			kbd_error(EXIT_FAILURE, 0, _("Couldn't get a file descriptor referring to the console."));

		/* check whether the keyboard is in Unicode mode */
		if (ioctl(fd, KDGKBMODE, &kbd_mode))
			kbd_error(EXIT_FAILURE, errno, _("Unable to read keyboard mode"));
		if (ioctl(fd, KDGETMODE, &kd_mode))
			kbd_error(EXIT_FAILURE, errno, _("Unable to read console mode"));

		if (kbd_mode == K_UNICODE) {
			if (options & OPT_A) {
				kbd_warning(0,
				        _("Warning: loading non-Unicode keymap on Unicode console\n"
				          "    (perhaps you want to do `kbd_mode -a'?)"));
			} else {
				flags |= LK_FLAG_PREFER_UNICODE;
			}

			/* reset -u option if keyboard is in K_UNICODE anyway */
			flags ^= LK_FLAG_UNICODE_MODE;

		} else if (options & OPT_U && kd_mode != KD_GRAPHICS) {
			kbd_warning(0,
			        _("Warning: loading Unicode keymap on non-Unicode console\n"
			          "    (perhaps you want to do `kbd_mode -u'?)"));
		}
	}

	lk_set_parser_flags(ctx, flags);

	dirpath = dirpath1;
	if ((ev = getenv("LOADKEYS_KEYMAP_PATH")) != NULL) {
		dirpath2[0] = ev;
		dirpath     = dirpath2;
	}

	if (options & OPT_D) {
		if (!(fp = kbdfile_new(fctx)))
			kbd_error(EXIT_FAILURE, 0, _("Unable to create kbdfile instance: %m"));

		/* first read default map - search starts in . */
		if (kbdfile_find(DEFMAP, dirpath, suffixes, fp))
			kbd_error(EXIT_FAILURE, 0, _("Unable to find file: %s"), DEFMAP);

		rc = lk_parse_keymap(ctx, fp);
		kbdfile_free(fp);

		if (rc == -1)
			goto fail;

	} else if (optind == argc) {
		if (!(fp = kbdfile_new(fctx)))
			kbd_error(EXIT_FAILURE, 0, _("Unable to create kbdfile instance: %m"));

		kbdfile_set_file(fp, stdin);
		kbdfile_set_pathname(fp, "<stdin>");

		rc = lk_parse_keymap(ctx, fp);
		kbdfile_free(fp);

		if (rc == -1)
			goto fail;
	}

	for (i = optind; argv[i]; i++) {
		if (!(fp = kbdfile_new(fctx))) {
			kbd_error(EXIT_FAILURE, 0, _("Unable to create kbdfile instance: %m"));
		}

		if (!strcmp(argv[i], "-")) {
			kbdfile_set_file(fp, stdin);
			kbdfile_set_pathname(fp, "<stdin>");

		} else if (kbdfile_find(argv[i], dirpath, suffixes, fp)) {
			kbd_warning(0, _("Unable to open file: %s: %m"), argv[i]);
			goto fail;
		}

		rc = lk_parse_keymap(ctx, fp);
		kbdfile_free(fp);

		if (rc == -1)
			goto fail;
	}

	if (!(options & OPT_P)) {
		if (options & OPT_B) {
			rc = lk_dump_bkeymap(ctx, stdout);
		} else if (options & OPT_M) {
			rc = lk_dump_ctable(ctx, stdout);
		} else {
			rc = lk_load_keymap(ctx, fd, kbd_mode);
		}
	}

fail:
	lk_free(ctx);
	kbdfile_context_free(fctx);

	if (fd >= 0)
		close(fd);

	if (rc < 0)
		exit(EXIT_FAILURE);

	exit(EXIT_SUCCESS);
}
